aiida_raspa/workchains/aiida_base.py

```diff
@@ -0,0 +1,437 @@
+"""
+RASPA Base WorkChain for Widom particle insertion, GCMC, and GEMC
+"""
+from __future__ import print_function
+from __future__ import absolute_import
+from copy import deepcopy
+
+from aiida.common import AttributeDict
+from aiida.plugins import CalculationFactory, DataFactory
+from aiida.orm import Code, Dict, Float, Int
+from aiida.engine import submit
+from aiida.engine import ToContext, WorkChain, if_, while_
+
+from aiida_raspa.utils.multiply_unitcell import multiply_unit_cell
+
+CifData = DataFactory("cif")
+ParameterData = DataFactory("dict")
+SinglefileData = DataFactory("singlefile")
+FolderData = DataFactory('folder')
+RaspaCalculation = CalculationFactory("raspa")
+
+# prepare_process_inputs and wrap_bare_dict_inputs are taken from aiida-cp2k plugin:
+# https://github.com/aiidateam/aiida-cp2k/blob/develop/aiida_cp2k/workchains/aiida_base_restart.py
+def prepare_process_inputs(process, inputs):
+    """Prepare the inputs for submission for the given process, according to its spec.
+    That is to say that when an input is found in the inputs that corresponds to an input port in the spec of the
+    process that expects a `Dict`, yet the value in the inputs is a plain dictionary, the value will be wrapped in by
+    the `Dict` class to create a valid input.
+    :param process: sub class of `Process` for which to prepare the inputs dictionary
+    :param inputs: a dictionary of inputs intended for submission of the process
+    :return: a dictionary with all bare dictionaries wrapped in `Dict` if dictated by the process spec
+    """
+    prepared_inputs = wrap_bare_dict_inputs(process.spec().inputs, inputs)
+    return AttributeDict(prepared_inputs)
+
+def wrap_bare_dict_inputs(port_namespace, inputs):
+    """Wrap bare dictionaries in `inputs` in a `Dict` node if dictated by the corresponding port in given namespace.
+    :param port_namespace: a `PortNamespace`
+    :param inputs: a dictionary of inputs intended for submission of the process
+    :return: a dictionary with all bare dictionaries wrapped in `Dict` if dictated by the port namespace
+    """
+    from aiida.engine.processes import PortNamespace
+
+    wrapped = {}
+
+    for key, value in inputs.items():
+
+        if key not in port_namespace:
+            wrapped[key] = value
+            continue
+
+        port = port_namespace[key]
+
+        if isinstance(port, PortNamespace):
+            wrapped[key] = wrap_bare_dict_inputs(port, value)
+        elif port.valid_type == Dict and isinstance(value, dict):
+            wrapped[key] = Dict(dict=value)
+        else:
+            wrapped[key] = value
+
+    return wrapped
+
+def check_widom_convergence(output_widom=None, comp_list=None, structure_label=None, conv_threshold=None):
```

Comment:
I would suggest to put it into utils, same for `check_gcmc_convergence`, `check_gemc_convergence`, `check_gemc_box`.